### PR TITLE
Feature/grapeshot

### PR DESCRIPTION
**Pull request by killanaca**
This PR adds the logic to load scripts in parallel on the `DOMContext` and `dom-context-harness` and to set a timeout and the error logic (if the script is mandatory or not) on each script.
Moreover add the Grapeshot logic to filter some kind of ads on some articles.

### CODE REVIEW

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/f0ec67dfbe344eb1507badbcd993ab3249349ecd/dextrose-test/index.html

File: packages/ad/__tests__/dom-context-harness.test.js [hunk 1]
```diff
@@ -129,45 +137,45 @@ describe("DOMContext harness", () => {
...
+      globalNames: ["requiredVar"],
+      eventCallback
     });
 
     harness.execute();
-
-    fireLoadEventFor("first");
-
-    expect(init).toHaveBeenCalledTimes(1);
+    expect(init).toHaveBeenCalledTimes(0);
```
Comment (colincclark) [inline packages/ad/__tests__/dom-context-harness.test.js]: `expect(init).not.toBeCalled();`

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/f4baa717d49725955356b8204a07e683c4adeedd/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/ef7bd654c00672d6969b485eefd4adb2304f8795/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/21e85361be19709af9444d4438715bcdf5b452da/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/6db279af23222fc4fe6a163cff66dae0695dc9b8/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/662b0ed59e7be59ed0f448b32f7f30ae4baf7163/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/5c7db5a228dac83c5b1c48836321b12adf8e9da5/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/80050afd57a765eb60cd13bddc58688785abb43a/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/267c59019bd52c5dce8d62d8611901319c2ee5d7/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/3213f02715130e0a74f8ef240f129db917953b7e/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/df48903a62c39cd7c91bb89b80c11b9b72ec3eba/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/44fa65e198b9a9000cf21da8ea44650ae30df4b6/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/bc55055f1e0e721d59cd547288bbf627dcf0de19/dextrose-test/index.html

File: packages/ad/__tests__/dom-context-harness.test.js [hunk 2]
```diff
@@ -104,68 +121,63 @@ describe("DOMContext harness", () => {
...
     harness.execute();
+    expect(init).not.toBeCalled();
+    jest.runAllTimers();
     expect(init).toHaveBeenCalledTimes(1);
   });
 
-  it("does not invoke the init function twice when multiple scripts load after globals are provided", () => {
-    window.requiredVar = "present";
+  it("invokes init function if the script can fail", () => {
+    jest.useFakeTimers();
```
Comment (craigbilner) [inline packages/ad/__tests__/dom-context-harness.test.js]: this should come from `jest-configurator` now

File: packages/ad/ad-init.js [hunk 3]
```diff
@@ -3,11 +3,20 @@
 
 /* eslint-env browser */
 const adInit = args => {
-  const { el, data, window, globals: { googletag }, renderComplete } = args;
+  const {
+    el,
+    data,
+    window,
+    globals: { googletag, gs_channels = "DEFAULT" }, // eslint-disable-line camelcase
```
Comment (craigbilner) [inline packages/ad/ad-init.js]: could this be `["DEFAULT"]`

File: packages/ad/dom-context-harness.js [hunk 4]
```diff
@@ -27,53 +29,109 @@ const makeHarness = ({
       eventCallback("error", e.message);
     }
   };
-
+  const log = (...message) => {
```
Comment (craigbilner) [inline packages/ad/dom-context-harness.js]: the message comes in parts?

File: packages/ad/dom-context-harness.js [hunk 5]
```diff
@@ -27,53 +29,109 @@ const makeHarness = ({
...
+    eventCallback("log", `${data.pos} ad: ${message.join(" ")}`);
+  };
   return {
     execute() {
       withCatch(() => {
-        this.injectScripts();
+        // eslint-disable-next-line no-param-reassign
+        window.scritpsProcessed = window.scritpsProcessed || [];
```
Comment (craigbilner) [inline packages/ad/dom-context-harness.js]: `scripts`

File: packages/ad/dom-context-harness.js [hunk 6]
```diff
@@ -27,53 +29,109 @@ const makeHarness = ({
...
+        this.addProcessedScript(scriptId);
+        this.runInitIfGlobalsPresent();
+      });
+    },
+    scriptErrored(scriptId, canRequestFail, err) {
+      withCatch(() => {
+        log("script errored:", scriptId);
+        this.addProcessedScript(scriptId);
+        if (!canRequestFail) {
+          throw new Error(`Failed to load the script ${err}`);
```
Comment (craigbilner) [inline packages/ad/dom-context-harness.js]: what does this mean for the user experience?

File: packages/ad/dom-context.js [hunk 7]
```diff
@@ -27,13 +27,15 @@ export default class DOMContext extends PureComponent {
       // don't try and process postMessage events from 3rd party scripts
       return;
     }
-    const event = JSON.parse(json);
-    if (event.type === "error" && process.env.NODE_ENV !== "production") {
+    const { type, detail } = JSON.parse(json);
```
Comment (craigbilner) [inline packages/ad/dom-context.js]: can this fail?

File: packages/ad/dom-context-harness.js (hunk 4 above)
Comment (killanaca) [reply packages/ad/dom-context-harness.js]: the idea is to have the same `parameters` like `console.log`

File: packages/ad/ad-init.js (hunk 3 above)
Comment (killanaca) [reply packages/ad/ad-init.js]: There shouldn't be an issue, but to double check with ads devs

File: packages/ad/dom-context.js (hunk 7 above)
Comment (killanaca) [reply packages/ad/dom-context.js]: Would be a good idea to revisit this bit after the prebidding functionality. I'll add a note.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/c1b4cbb6449163349a46725778c8ca3d748af70b/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/1d9174e6a18b367fd2638874bfe551b70950b6a9/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/5792dbbff35d3eefabcfcc23e59c5557989e01e4/dextrose-test/index.html

**Pull request merged**